test-poc/framework/src/main/java/org/keycloak/test/framework/injection/Supplier.java

```diff
@@ -1,26 +1,39 @@
 package org.keycloak.test.framework.injection;
 
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Optional;
 
 public interface Supplier<T, S extends Annotation> {
 
     Class<S> getAnnotationClass();
 
     Class<T> getValueType();
 
-    InstanceWrapper<T, S> getValue(Registry registry, S annotation);
-
-    default InstanceWrapper<T, S> getValue(Registry registry, S annotation, Class<? extends T> valueType) {
-        return getValue(registry, annotation);
+    T getValue(InstanceContext<T, S> instanceContext);
+
+    default LifeCycle getLifeCycle(S annotation) {
```

Comment:
I'd do:

```java
if (annotation == null) {
  return getDefaultLifecycle();
}

// The rest
```

Reply:
Don't think it matters too much - down to personal preference I guess ;)

Reply:
Oh for sure, you know how much I hate nesting, this is just a prime example of an [early return](https://medium.com/swlh/return-early-pattern-3d18a41bba8) pattern.